# Decision Trees and Ensemble Learning

## Credit risk scoring project

**Dataset**

In [1]:
!wget https://github.com/gastonstat/CreditScoring/raw/master/CreditScoring.csv

--2022-10-11 11:32:11--  https://github.com/gastonstat/CreditScoring/raw/master/CreditScoring.csv
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/gastonstat/CreditScoring/master/CreditScoring.csv [following]
--2022-10-11 11:32:11--  https://raw.githubusercontent.com/gastonstat/CreditScoring/master/CreditScoring.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 182489 (178K) [text/plain]
Saving to: ‘CreditScoring.csv.7’

CreditScoring.csv.7 100%[===================>] 178.21K  --.-KB/s    in 0.02s   

2022-10-11 11:32:11 (7.37 MB/s) - ‘CreditScoring.csv.7’ saved [182489/182489]



### Librairies

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import missingno as msno

plt.style.use('ggplot')

**Data cleanining**

In [3]:
df = pd.read_csv('CreditScoring.csv')
df.head()

Status  Seniority  Home  Time  Age  Marital  Records  Job  Expenses  \
0       1          9     1    60   30        2        1    3        73   
1       1         17     1    60   58        3        1    1        48   
2       2         10     2    36   46        2        2    3        90   
3       1          0     1    60   24        1        1    1        63   
4       1          0     1    36   26        1        1    1        46   

   Income  Assets  Debt  Amount  Price  
0     129       0     0     800    846  
1     131       0     0    1000   1658  
2     200    3000     0    2000   2985  
3     182    2500     0     900   1325  
4     107       0     0     310    910

In [4]:
# make all columns in lower
df.columns = df.columns.str.lower()

In [16]:
df.shape

(4455, 14)

**Mapping the target**

In [5]:
df['status'].value_counts(dropna=False)

1    3200
2    1254
0       1
Name: status, dtype: int64

In [6]:
df['status'] = df['status'].map({
    1:'ok',
    2:'default',
    0:'unk'
})

** Mapping categorical features **

In [7]:
def mapping_categorical(df, cat, cat_lst):
  to_lst = df[cat].value_counts().sort_index().index.to_list()
  cat_lst = cat_lst

  df[cat] = (
      df[cat].map({
          k:v for (k,v) in zip(to_lst, cat_lst)
      })
   )

In [8]:
cols = ['home', 'marital', 'records', 'job']

home_lst = ['unk', 'rent', 'owner', 'private', 'ignore', 'parents', 'other']
marital_lst = ['unk', 'single', 'married', 'widow', 'separated', 'divorced']
records_lst = ['no', 'yes', 'unk']
job_lst = ['unk', 'fixed', 'partime', 'freelance', 'others']
cat_lst = [home_lst, marital_lst, records_lst, job_lst]

for col, cat in zip(cols, cat_lst):
  mapping_categorical(df, col, cat)


In [9]:
df.head()

status  seniority   home  time  age  marital records        job  expenses  \
0       ok          9   rent    60   30  married      no  freelance        73   
1       ok         17   rent    60   58    widow      no      fixed        48   
2  default         10  owner    36   46  married     yes  freelance        90   
3       ok          0   rent    60   24   single      no      fixed        63   
4       ok          0   rent    36   26   single      no      fixed        46   

   income  assets  debt  amount  price  
0     129       0     0     800    846  
1     131       0     0    1000   1658  
2     200    3000     0    2000   2985  
3     182    2500     0     900   1325  
4     107       0     0     310    910

**Check numerical values**

In [11]:
df.describe().T.round()

count       mean         std    min     25%     50%     75%  \
seniority  4455.0        8.0         8.0    0.0     2.0     5.0    12.0   
time       4455.0       46.0        15.0    6.0    36.0    48.0    60.0   
age        4455.0       37.0        11.0   18.0    28.0    36.0    45.0   
expenses   4455.0       56.0        20.0   35.0    35.0    51.0    72.0   
income     4455.0   763317.0   8703625.0    0.0    80.0   120.0   166.0   
assets     4455.0  1060341.0  10217569.0    0.0     0.0  3500.0  6000.0   
debt       4455.0   404382.0   6344253.0    0.0     0.0     0.0     0.0   
amount     4455.0     1039.0       475.0  100.0   700.0  1000.0  1300.0   
price      4455.0     1463.0       628.0  105.0  1118.0  1400.0  1692.0   

                  max  
seniority        48.0  
time             72.0  
age              68.0  
expenses        180.0  
income     99999999.0  
assets     99999999.0  
debt       99999999.0  
amount         5000.0  
price         11140.0

One thing we immediately can notice is that the max value is 99999999 in some cases. This is quite suspicious. As it turns out, it’s an artificial value — this is how missing values are encoded in this dataset.

In [12]:
def fix_missing_values(df, val_to_rep, rep, *f_lst):
  for f in f_lst:
    df[f] = df[f].replace(val_to_rep, rep)

In [13]:
fix_missing_values(df, 99999999.0, np.nan, ['income', 'assets', 'debt'])

In [14]:
df.describe().T.round()

count    mean      std    min     25%     50%     75%       max
seniority  4455.0     8.0      8.0    0.0     2.0     5.0    12.0      48.0
time       4455.0    46.0     15.0    6.0    36.0    48.0    60.0      72.0
age        4455.0    37.0     11.0   18.0    28.0    36.0    45.0      68.0
expenses   4455.0    56.0     20.0   35.0    35.0    51.0    72.0     180.0
income     4421.0   131.0     86.0    0.0    80.0   120.0   165.0     959.0
assets     4408.0  5403.0  11573.0    0.0     0.0  3000.0  6000.0  300000.0
debt       4437.0   343.0   1246.0    0.0     0.0     0.0     0.0   30000.0
amount     4455.0  1039.0    475.0  100.0   700.0  1000.0  1300.0    5000.0
price      4455.0  1463.0    628.0  105.0  1118.0  1400.0  1692.0   11140.0

In [15]:
df.isna().sum()

status        0
seniority     0
home          0
time          0
age           0
marital       0
records       0
job           0
expenses      0
income       34
assets       47
debt         18
amount        0
price         0
dtype: int64

We notice that there’s one row with “unknown” status: we don’t know whether this client
managed to pay back the loan or not. For our project, this row is not useful

In [17]:
df = df[df.status != 'unk']